src/System.Private.CoreLib/System.Private.CoreLib.csproj

```
@@ -146,6 +146,7 @@
     <Compile Include="$(BclSourcesRoot)\System\Diagnostics\Eventing\EventPipeEventProvider.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Diagnostics\Eventing\EventPipeMetadataGenerator.cs" />
     <Compile Include="$(BclSourcesRoot)\System\Diagnostics\Eventing\EventPipePayloadDecoder.cs" />
+    <Compile Include="$(BclSourcesRoot)\System\Diagnostics\Eventing\RuntimeEventSource.cs" Condition="'$(FeaturePerfTracing)' == 'true'"/>
```

I think we need to do a little more to prevent partially finished work from leaking to customers. Although this type has been defined as internal, I don't see anything that would prevent an instance of this type from being handed to any customer that implements EventListener.OnEventSourceCreated. There are some other APIs that would interact with it indirectly too. I'd suggest using a new feature flag that will not be true by default until we believe the work is customer-ready.

```suggestion
    <Compile Include="$(BclSourcesRoot)\System\Diagnostics\Eventing\RuntimeEventSource.cs" Condition="'$(FeaturePerfCounters)' == 'true'"/>
```

---

@noahfalk Do you expect this to be finished for .NET Core 3.0? If yes, then you do not need this feature flag. It is ok to ship unpolished features in previews. If not, then this should be developed in a branch as a long term project probably.

---

I do expect it to be done prior to 3.0 RTM, but I didn't have much clarity on whether there would be earlier preview releases that still have a high quality bar (a go-live preview for example). My other concern is that we might get some customers that wind up using this code without any express intention to do so. For example any app that enables/disables all providers using EventListener. I don't think the risk is particularly high, so if you think that matches the level of caution other areas of the product are taking in previews I'm happy to follow suit.

---

When in doubt, do the simpler thing.   I sounds like we can avoid the complexity of creating a feature flag.   I think this is fine as formally we don't guarantee compat in previews, and practically no one is going to care (turning on all EventSource is generally not done (some are quite high volume), and now one will know about the one we are adding here unless we tell them.